src/fsharp/NameResolution.fs

```diff
@@ -3748,11 +3749,29 @@ let rec ResolvePartialLongIdentInModuleOrNamespace (ncenv: NameResolver) nenv is
              mty.TypeDefinitions
              |> List.filter (fun tcref -> not (tcref.LogicalName.Contains(",")))
              |> List.filter (fun tycon -> not (IsTyconUnseen ad g ncenv.amap m (modref.NestedTyconRef tycon)))
-
-         let ilTyconNames = 
-             mty.TypesByAccessNames.Values
-             |> List.choose (fun (tycon:Tycon) -> if tycon.IsILTycon then Some tycon.DisplayName else None)
-             |> Set.ofList
+         
+         let accessibleSubModules =
+             let moduleOrNamespaces =
+                 mty.ModulesAndNamespacesByDemangledName
+                 |> NameMap.range
+
+             if isNil moduleOrNamespaces then [] else
+
+             let ilTyconNames =
+                 let hashSet = HashSet()
+                 for tycon in mty.TypesByAccessNames.Values do
+                    if tycon.IsILTycon then 
+                        hashSet.Add tycon.DisplayName |> ignore
+                 hashSet
+
+             moduleOrNamespaces
+             |> List.filter (fun x -> 
+                 let demangledName = x.DemangledModuleOrNamespaceName
+                 notFakeContainerModule ilTyconNames demangledName && IsInterestingModuleName demangledName)
+             |> List.map modref.NestedTyconRef
+             |> List.filter (IsTyconUnseen ad g ncenv.amap m >> not)
+             |> List.filter (EntityRefContainsSomethingAccessible ncenv m ad)
```

Comment:
Can these checks be combined to a single List.filter? Why create temporary lists to just do a check?

Reply:
this could be done in yet another PR